Document how to debug performance of Prow jobs in etcd (#20796)
etcd-io/etcd: Documentation/contributor-guide/prow_jobs.md

# Prow Jobs in etcd

## 1. Introduction to Prow

[Prow](https://docs.prow.k8s.io/docs/) is a Kubernetes based CI/CD system. Jobs can be triggered by various types of events and report their status to many different services. Prow provides GitHub automation through policy enforcement and chat-ops via `/command` interactions on pull requests (e.g., `/test`, `/approve`, `/retest`), enabling contributors to trigger jobs and manage workflows directly from GitHub comments.

When a user comments `/ok-to-test`or `/retest,` on a Pull Request, GitHub sends a webhook to Prow's Kubernetes cluster. Visit this [site](https://docs.prow.k8s.io/docs/life-of-a-prow-job/) to further understand the lifecycle of a Prow job.
This is where you can find all etcd Prow jobs [status](https://prow.k8s.io/?repo=etcd-io%2Fetcd)

## 2. How Prow is used for etcd Testing

> The CI of etcd is managed by kubernetes/test-infra, which leverages prow inside it.

Reviewer: ```suggestion
etcd's CI is managed by [kubernetes/test-infra](https://github.com/kubernetes/test-infra), running Prow.
```

- Robustness:
  - Description: Long-running, fault-injection and chaos-style end-to-end tests that validate etcd correctness and availability under failures (node crashes, network partitions, resource exhaustion, upgrades).
  - When to run: Periodics for continuous coverage; run for PRs that touch consensus, storage, recovery, or upgrade paths.
  - Example job patterns: pull-etcd-robustness, periodic-robustness

- Integration:
  - Description: Functional end-to-end and cross-component tests that exercise real client/server interactions, snapshots/restore, upgrades and compatibility across OS/arch.
  - When to run: Presubmits for PRs that change APIs, client behavior, or integration points; periodics for broad platform coverage.
  - Example job patterns: pull-etcd-e2e-amd64, pull-etcd-integration

- Static check:
  - Description: Fast, deterministic checks (build, unit tests, linters, go vet/staticcheck, formatting, license/header checks, generated-code verification) that catch style, correctness and packaging problems early.
  - When to run: Every PR as presubmits; quick feedback loop before running expensive tests.

>   - Example job patterns: pull-etcd-verify, pull-etcd-lint, pull-etcd-unit

Author: Alright. I have reordered the structure. Let me know what you think and I can adjust the restructure.

> # Prow Jobs in etcd

Reviewer 2: This document isn't really about prow jobs in general, but is mostly about looking at prow job resource usage.  Should the name (and commit message) be updated to reflect that?

When a user comments `/ok-to-test`or `/retest,` on a Pull Request, GitHub sends a webhook to Prow's Kubernetes cluster. Visit this [site](https://docs.prow.k8s.io/docs/life-of-a-prow-job/) to further understand the lifecycle of a Prow job.
This is where you can find all etcd Prow jobs [status](https://prow.k8s.io/?repo=etcd-io%2Fetcd)

## 2. How Prow is used for etcd Testing

etcd's CI is managed by [kubernetes/test-infra](https://github.com/kubernetes/test-infra), running Prow.

When a pull request is submitted, or a `/command` is issued, the CI of etcd which managed by [kubernetes/test-infra](https://github.com/kubernetes/test-infra) uses Prow to run the tests. You can view all supported Prow [commands](https://prow.k8s.io/command-help).

### Jobs Types

The jobs [configuration](https://github.com/kubernetes/test-infra/tree/master/config/jobs/etcd) for etcd.

There are 3 different job types:

- Presubmits run against code in PRs
- Postsubmits run after merging code

> - Periodics run on a periodic basis

Reviewer 2: All three of these can link to Etcd's config for each.

Please see [ProwJob](https://docs.prow.k8s.io/docs/jobs/) docs for more info.

As an [example](https://github.com/kubernetes/test-infra/blob/master/config/jobs/etcd/etcd-presubmits.yaml), are the presubmits jobs of etcd. `pull-etcd-e2e-amd64` is one of the [presubmits](https://github.com/kubernetes/test-infra/blob/b21a1d3a72d5715ea7c9234cade21751847cfbe5/config/jobs/etcd/etcd-presubmits.yaml#L193).

Refer to [the test-infra Job Types documentation](https://github.com/kubernetes/test-infra/tree/master/config/jobs#job-types) to learn more about them.

### How to manually run a given job on Prow

> These tests can be triggered when you leave a comment, like `/ok-to-test` (only triggered by an etcd-io member) or `/retest`, in PR [example](https://github.com/etcd-io/etcd/pull/20733#issuecomment-3341443205). You can find all supported [commands](https://prow.k8s.io/command-help).

Reviewer 2: As a future doc PR, you should look up how to run a specific test.

If the “requested” line is much higher than “used”, it suggests the job’s request could be tuned downward.

- CPU panel

Similar structure: green = actual usage, orange/yellow = requested CPU, red = CPU limit (if set).
Y‑axis often in number of CPU cores or fraction thereof (e.g., 1.0 = one core).
A green line with spikes may show bursts of CPU usage (e.g., build or compile phases) while idle periods show low usage.
If CPU usage consistently saturates the limit, the job may be throttled or delayed. If usage is consistently far below request, tuning may reduce cost.

It is useful for a few reasons:

1. Tuning resources: By drilling into each build-run, you can determine realistic memory & CPU requests and limits for that job‑type. This helps avoid waste or avoid failed builds hitting resource limits.

2. Spotting anomalies: If one build suddenly used 8 GiB while normally this job uses 1 GiB, it may indicate a regression or mis‑configuration.

3. Capacity planning: Seeing typical and peak usage helps cluster operators plan node sizes, scheduling, concurrency of builds, etc.

> 4. Debugging performance issues: A build with unexpectedly high CPU or memory might be stuck, looping, or consuming resources inefficiently.

Reviewer 2: Move all this to the top of the section; Navigating Performance should start with why it's useful, and then move on to the individual panels, rather than the current flow.

# Prow Jobs in etcd

## 1. Introduction to Prow

> [Prow](https://docs.prow.k8s.io/docs/) is a Kubernetes based CI/CD system. Jobs can be triggered by various types of events and report their status to many different services. In addition to job execution, Prow provides GitHub automation in the form of policy enforcement, chat-ops via `/foo` style commands, and automatic PR merging.

Reviewer: I would emphasize that it responds to pull request comments using `/commands`.

> When a user comments `/ok-to-test`or `/retest,` on a Pull Request, GitHub sends a webhook to Prow's Kubernetes cluster. The request travels through an ingress for TLS termination, gets routed to the hook service, and arrives at the hook application running in pods. Visit this [site](https://docs.prow.k8s.io/docs/life-of-a-prow-job/) to further understand the lifecycle of a Prow job.

Author: Sure. It is removed. 
I was trying to what happens behing a prow command.

This is where you can find all etcd Prow jobs [status](https://prow.k8s.io/?repo=etcd-io%2Fetcd)

## 2. How Prow is used for etcd Testing

The CI of etcd is managed by kubernetes/test-infra, which leverages prow inside it.

> Whenever a pull request is submitted, or a command is called, the CI of etcd, managed by [kubernetes/test-infra](https://github.com/kubernetes/test-infra) leverages Prow to run the tests. You can find all the supported [commands](https://prow.k8s.io/command-help).

Author: Done

> The [history](https://prow.k8s.io/job-history/gs/kubernetes-ci-logs/pr-logs/directory/pull-etcd-e2e-amd64?buildId=)of the ran can be found.

Reviewer: This paragraph looks unfinished, right?

Author: Done, omitted it

### Jobs Types

The jobs [configuration](https://github.com/kubernetes/test-infra/tree/master/config/jobs/etcd) for etcd.

There are 3 different job types:

- Presubmits run against code in PRs
- Postsubmits run after merging code
- Periodics run on a periodic basis

Please see [ProwJob](https://docs.prow.k8s.io/docs/jobs/) docs for more info.

As an [example](https://github.com/kubernetes/test-infra/blob/master/config/jobs/etcd/etcd-presubmits.yaml), are the presubmits jobs of etcd. `pull-etcd-e2e-amd64` is one of the [presubmits](https://github.com/kubernetes/test-infra/blob/b21a1d3a72d5715ea7c9234cade21751847cfbe5/config/jobs/etcd/etcd-presubmits.yaml#L193).

Refer to [this](https://github.com/kubernetes/test-infra/tree/master/config/jobs#job-types) to learn more about job types.

### How to manually run a given job on Prow

> These tests can be triggered when you leave a comment, like `/ok-to-test` or `/retest`, in PR [example](https://github.com/etcd-io/etcd/pull/20733#issuecomment-3341443205).

Reviewer: Noting that `/ok-to-test` can only be triggered by an etcd-io member.

You can find all supported [commands](https://prow.k8s.io/command-help).

## 3. Navigating Performance Dashboard (Grafana)

> One example of [dashboards](https://monitoring-gke.prow.k8s.io/d/96Q8oOOZk/builds?orgId=1&refresh=30s&var-org=etcd-io&var-repo=etcd&var-job=ci-etcd-robustness-main-amd64&var-build=All&from=now-7d&to=now) can be found.

Reviewer: We should clearly list the gke and eks Grafana dashboards, not just as examples.
